core/client/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -117,54 +120,54 @@ protected FileInStream(URIStatus status, InStreamOptions options) {
     mFileLength = status.getLength();
     mContext = FileSystemContext.INSTANCE;
     mAlluxioStorageType = options.getAlluxioStorageType();
-    mShouldCacheCurrentBlock = mAlluxioStorageType.isStore();
+    mShouldCache = mAlluxioStorageType.isStore();
     mShouldCachePartiallyReadBlock = options.isCachePartiallyReadBlock();
     mClosed = false;
     mLocationPolicy = options.getLocationPolicy();
-    if (mShouldCacheCurrentBlock) {
+    if (mShouldCache) {
       Preconditions.checkNotNull(options.getLocationPolicy(),
           PreconditionMessage.FILE_WRITE_LOCATION_POLICY_UNSPECIFIED);
     }
-    mSingleByte = new byte[1];
   }
 
   @Override
   public void close() throws IOException {
     if (mClosed) {
       return;
     }
-    if (mShouldCacheCurrentBlock && mShouldCachePartiallyReadBlock) {
-      readCurrentBlockToPos(mFileLength);
+    updateStreams();
+    if (mCurrentCacheStream != null && mShouldCachePartiallyReadBlock) {
+      readCurrentBlockToPos(Long.MAX_VALUE);
     }
     if (mCurrentBlockInStream != null) {
       mCurrentBlockInStream.close();
     }
-    closeCacheStream();
+    closeOrCancelCacheStream();
     mClosed = true;
   }
 
   @Override
   public int read() throws IOException {
-    if (!validPosition(mPos)) {
+    if (remaining() <= 0) {
       return -1;
     }
+    updateStreams();
+    Preconditions.checkState(mCurrentBlockInStream != null, "Reached EOF unexpectedly.");
 
-    checkAndAdvanceBlockInStream();
     int data = mCurrentBlockInStream.read();
+
+    // This should not happen?
```

Comment:
That was more of a question. I don't know how this can happen when remaining() > 0 (or in the old version when mPos < mFileLength). I have removed it